calico_upgrade/Makefile

```diff
@@ -0,0 +1,339 @@
+
+
+.PHONY: stop-etcd
+stop-etcd:
+	@-docker rm -f calico-etcd
+
+# This depends on clean to ensure that dependent images get untagged and repulled
+.PHONY: semaphore
+semaphore: clean
+	# Clean up unwanted files to free disk space.
+	bash -c 'rm -rf /home/runner/{.npm,.phpbrew,.phpunit,.kerl,.kiex,.lein,.nvm,.npm,.phpbrew,.rbenv} /usr/local/golang /var/lib/mongodb'
+
+	# Run the containerized tests first.
+	$(MAKE) test-containerized st
+
+	$(MAKE) calico/upgrade
+
+	# Make sure that calico-upgrade builds cross-platform.
+	$(MAKE) dist/calico-upgrade-darwin-amd64 dist/calico-upgrade-windows-amd64.exe
+
+release: clean
+ifndef VERSION
+	$(error VERSION is undefined - run using make release VERSION=vX.Y.Z)
+endif
+	git tag $(VERSION)
+
+	# Check to make sure the tag isn't "-dirty".
+	if git describe --tags --dirty | grep dirty; \
+	then echo current git working tree is "dirty". Make sure you do not have any uncommitted changes ;false; fi
+
+	# Build the calico-upgrade binaries, as well as the calico/upgrade image.
+	$(MAKE) dist/calico-upgrade dist/calico-upgrade-darwin-amd64 dist/calico-upgrade-windows-amd64.exe
+	$(MAKE) calico/upgrade
+
+	# Check that the version output includes the version specified.
+	# Tests that the "git tag" makes it into the binaries. Main point is to catch "-dirty" builds
+	# Release is currently supported on darwin / linux only.
+	if ! docker run $(UPGRADE_CONTAINER_NAME) version | grep 'Version:\s*$(VERSION)$$'; then \
+	  echo "Reported version:" `docker run $(UPGRADE_CONTAINER_NAME) version` "\nExpected version: $(VERSION)"; \
+	  false; \
+	else \
+	  echo "Version check passed\n"; \
+	fi
+
+	# Retag images with corect version and quay
+	docker tag $(UPGRADE_CONTAINER_NAME) $(UPGRADE_CONTAINER_NAME):$(VERSION)
+	docker tag $(UPGRADE_CONTAINER_NAME) quay.io/$(UPGRADE_CONTAINER_NAME):$(VERSION)
+	docker tag $(UPGRADE_CONTAINER_NAME) quay.io/$(UPGRADE_CONTAINER_NAME):latest
+
+	# Check that images were created recently and that the IDs of the versioned and latest images match
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" $(UPGRADE_CONTAINER_NAME)
+	@docker images --format "{{.CreatedAt}}\tID:{{.ID}}\t{{.Repository}}:{{.Tag}}" $(UPGRADE_CONTAINER_NAME):$(VERSION)
+
+	@echo "\nNow push the tag and images. Then create a release on Github and"
+	@echo "attach dist/calico-upgrade, dist/calico-upgrade-darwin-amd64, and dist/calico-upgrade-windows-amd64.exe binaries"
+	@echo "\nAdd release notes for calico-upgrade. Use this command"
+	@echo "to find commit messages for this release: git log --oneline <old_release_version>...$(VERSION)"
+	@echo "git push origin $(VERSION)"
+	@echo "docker push $(UPGRADE_CONTAINER_NAME):$(VERSION)"
+	@echo "docker push quay.io/$(UPGRADE_CONTAINER_NAME):$(VERSION)"
+	@echo "docker push $(UPGRADE_CONTAINER_NAME):latest"
+	@echo "docker push quay.io/$(UPGRADE_CONTAINER_NAME):latest"
+	@echo "See RELEASING.md for detailed instructions."
```

Comment:
someone's going to write a RELEASING.md for calico_upgrade?

Reply:
ha!  good point.  I shall remove the merest suggestion of such a thing.